clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -493,6 +493,148 @@ class CIRFuncLowering : public mlir::OpConversionPattern<mlir::cir::FuncOp> {
+  auto type = attr.getType().cast<mlir::cir::ArrayType>().getEltType();
+  auto values = llvm::SmallVector<T, 8>{};
+  for (auto element : attr.getValue().cast<mlir::ArrayAttr>())
+    values.push_back(element.cast<mlir::IntegerAttr>().getInt());
+  return mlir::DenseElementsAttr::get(
+      mlir::RankedTensorType::get({(int64_t)values.size()}, type),
+      llvm::ArrayRef(values));
+}
+
+mlir::LLVM::Linkage convertLinkage(mlir::cir::GlobalLinkageKind linkage) {
+  using CIR = mlir::cir::GlobalLinkageKind;
+  using LLVM = mlir::LLVM::Linkage;
+
+  switch (linkage) {
+  case CIR::AvailableExternallyLinkage:
+    return LLVM::AvailableExternally;
+  case CIR::CommonLinkage:
+    return LLVM::Common;
+  case CIR::ExternalLinkage:
+    return LLVM::External;
+  case CIR::ExternalWeakLinkage:
+    return LLVM::ExternWeak;
+  case CIR::InternalLinkage:
+    return LLVM::Internal;
+  case CIR::LinkOnceAnyLinkage:
+    return LLVM::Linkonce;
+  case CIR::LinkOnceODRLinkage:
+    return LLVM::LinkonceODR;
+  case CIR::PrivateLinkage:
+    return LLVM::Private;
+  case CIR::WeakAnyLinkage:
+    return LLVM::Weak;
+  case CIR::WeakODRLinkage:
+    return LLVM::WeakODR;
+  };
+}
+
+class CIRGetGlobalOpLowering
+    : public mlir::OpConversionPattern<mlir::cir::GetGlobalOp> {
+public:
+  using OpConversionPattern<mlir::cir::GetGlobalOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::GetGlobalOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    auto type = getTypeConverter()->convertType(op.getType());
+    auto symbol = op.getName();
+    rewriter.replaceOpWithNewOp<mlir::LLVM::AddressOfOp>(op, type, symbol);
+    return mlir::success();
+  }
+};
+
+class CIRGlobalOpLowering
+    : public mlir::OpConversionPattern<mlir::cir::GlobalOp> {
+public:
+  using OpConversionPattern<mlir::cir::GlobalOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::GlobalOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+
+    // fetch required values to create LLVM op
+    auto type = getTypeConverter()->convertType(op.getSymType());
+    auto isConst = op.getConstant();
+    auto linkage = convertLinkage(op.getLinkage());
+    auto symbol = op.getSymName();
+    auto init = op.getInitialValue();
+
+    // check for missing funcionalities
+    if (!init.has_value()) {
+      op.emitError() << "uninitialized globals are not yet supported.";
+      return mlir::failure();
+    }
+
+    // initializer is a constant array: convert it to a compatible llvm init
+    if (auto constArr = dyn_cast<mlir::cir::ConstArrayAttr>(init.value())) {
+      if (auto attr = dyn_cast<mlir::StringAttr>(constArr.getValue())) {
+        init = rewriter.getStringAttr(attr.getValue());
+      } else if (auto attr = dyn_cast<mlir::ArrayAttr>(constArr.getValue())) {
```

Comment:
You can also do these dyn_casts on attributes this way: `constArr.getValue().dyn_cast<ArrayAttr>()`, which I think reads better - this is fine, just wanted to point it out the option.